Appc 1147 wallet's custom store page

mikieribeiro opened the pull request:
**What does this PR do?**

The call to the Wallet's store page was changed for this pull request. A verification is now being done to determine if Aptoide is installed on the user's device when he launches the billing flow
The wallet's store page is opened directly on Aptoide if it is installed or on any other market in case it's not
**Database changed?**

   No

**Where should the reviewer start?**

- [ ] GitHub/bds-sdk/android-appcoins- 
       billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java


**How should this be manually tested?**

  The toolbox should be ran and the launch billing flow started (the first wallet installation pop up from the poa must be disregarded). Test with and without Aptoide installed on the device and see if the wallet's page is opened directly on it or on the google play store in case aptoide ins't installed (the wallet package name must be "com.appcoins.wallet" for this test to take place)

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-1147](https://aptoide.atlassian.net/browse/APPC-1147)

**Questions:**

  How can we improve Aptoide's application starting time? It has a significant delay comparing to google play (opening the wallet's store page)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

mikieribeiro committed:
APPC-1147 Custom wallet's store page call

-Changed the call to Wallet's store page to be custom
-Aptoide's installation is verified when the billing flow is launched
-The wallet's store page is opened directly on aptoide if it is installed or on any other market in case it's not

mikieribeiro committed:
APPC-1147 Change call to Wallet's store page to be custom

-Fixed uncorrect changes to debug's buildConfigFields

mikieribeiro committed:
Appc-1147-Wallet's-Custom-Store-Page

-Optimized the promptToInstallWallet () method so no unnecessary code would be repeated

brunovarelaAptoide commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -46,24 +56,38 @@ public static void promptToInstallWallet() {
     }
+
+    final Intent appStoreIntent =
+        new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + walletPackageName));
```
is the walletPackageName needed?

brunovarelaAptoide commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -34,7 +35,16 @@ public static boolean hasWalletInstalled() {
 
+  public static boolean hasAptoideInstalled(String packageName, PackageManager packageManager) {
+    try {
+      return packageManager.getApplicationInfo(packageName, 0).enabled;
```
good idea check if the app is enabled.Kudos

brunovarelaAptoide commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -46,24 +56,38 @@ public static void promptToInstallWallet() {
+      builder.setPositiveButton(R.string.install, new DialogInterface.OnClickListener() {
+        @Override public void onClick(DialogInterface dialog, int which) {
+          try {
+            if (hasAptoideInstalled(aptoidePackageName, context.getPackageManager())) {
```
Does this function need does parameters?aptoidePackageName is a global variable and context is too.

brunovarelaAptoide replied:
does this if need to be inside of the try catch block?

brunovarelaAptoide commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -46,24 +56,38 @@ public static void promptToInstallWallet() {
-    });
+      builder.setPositiveButton(R.string.install, new DialogInterface.OnClickListener() {
+        @Override public void onClick(DialogInterface dialog, int which) {
+          try {
```
can the code inside of onClick be split in a different function for an easier code read?

mikieribeiro committed:
Appc-1147-Wallet's-Custom-Store-Page

-Made relevant changes to the aptoide installation verification methods and prompt to install wallet methods so the code would be more efficient and easier to understand

brunovarelaAptoide committed:
Changes to a billing final version

- Change the way Wallet utils gets the activity.

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingStubHelper.java:
```diff
@@ -186,6 +231,11 @@ private void createRepository() {
             lockThread.notify();
             isServiceBound = true;
+            if (isMainThread) {
+              Toast.makeText(context, "Try again, it will work this time =)", Toast.LENGTH_SHORT)
```
Why do we need this toast here?

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -46,47 +58,53 @@ public static void promptToInstallWallet() {
-    builder.setTitle(R.string.wallet_missing);
-    builder.setMessage(act.getString(R.string.install_wallet_from_iab));
-    Log.d("String name: ", act.getString(R.string.install_wallet_from_iab));
+    builder.setTitle("APPC Wallet Missing");
```
Why do we have hard coded string for the dialog content instead of string ids?

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -46,47 +58,53 @@ public static void promptToInstallWallet() {
     });
 
     builder.setIcon(android.R.drawable.ic_dialog_alert);
+    Log.d(TAG, "promptToInstallWallet: builder.show()");
```
do we need this log?

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -46,47 +58,53 @@ public static void promptToInstallWallet() {
+    //Check if the user has aptoide installed and open the aptoide's wallet page
+    if (hasAptoideInstalled()) {
+      Log.d("AptoideInstallation", "Aptoide is installed on this device");
+      appStoreIntent.setPackage(aptoidePackageName);
```
There as been an update and additional details are required for the dialog that redirects to Aptoide.

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -46,47 +58,53 @@ public static void promptToInstallWallet() {
+    if (hasAptoideInstalled()) {
+      Log.d("AptoideInstallation", "Aptoide is installed on this device");
+      appStoreIntent.setPackage(aptoidePackageName);
+    } else {
```
Please don't leave a else clause for the solo purpose of logging that Aptoide is not installed and usualy the tag of the log should be the name of the class as the TAG define in this same class.

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -2,29 +2,29 @@
 
+  private static final String TAG = AppcoinsBillingStubHelper.class.getSimpleName();
   public static String walletPackageName = BuildConfig.BDS_WALLET_PACKAGE_NAME;
+  public static String aptoidePackageName = BuildConfig.APTOIDE_PACKAGE_NAME;
```
Do we need a local variable for this parameters since we are going for the BuildConfigs, since this is an utility class with static para meters this should be provided in the method call. If the idea is to have an utility class that is configured with parameters, please consider having it defined as an standard class without static methods

mikieribeiro committed:
Merge branch 'dev' into APPC-1147-Wallet's-custom-store-page

mikieribeiro committed:
APPC-1147-Wallet-Custom-Store-Page-Call

-Made code review changes

Trinkes commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -45,30 +57,51 @@ public static void promptToInstallWallet() {
+    if (hasAptoideInstalled()) {
+      Log.d(TAG, "Aptoide is installed on this device");
+      appStoreIntent.setPackage(BuildConfig.APTOIDE_PACKAGE_NAME);
+      url += aptoideParameters;
```
you're not used the new url, the old intent is the one being used

Trinkes commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -45,30 +57,51 @@ public static void promptToInstallWallet() {
     builder.show();
   }
 
+  private static void OnInstallClickAction(Activity act, Intent appStoreIntent, String url) {
```
you should create the AppStoreIntent in one method, here you have it half of it created on promptToInstallWallet and the other half in OnInstallClickAction

mikieribeiro committed:
Made requested changes to the pull request (utm url)

-Utm url is now being used by both aptoide and googlePlay

mikieribeiro committed:
Merge remote-tracking branch 'remotes/origin/dev' into APPC-1147-Wallet's-custom-store-page

mikieribeiro committed:
Merged dev branch and updated wallet's custom page call

-Wallet's nstallation being redirect to aptoide in case the user has it installed

JoaoRaimundo commented on gradle.properties:
```diff
@@ -14,6 +14,7 @@ ROPSTEN_NETWORK_BACKEND_BASE_HOST=https://apichain-dev.blockchainds.com
 MAIN_NETWORK_BACKEND_BASE_HOST_WS=https://api.blockchainds.com
 ROPSTEN_NETWORK_BACKEND_BASE_HOST_WS=https://api-dev.blockchainds.com
 
+APTOIDE_PACKAGE_NAME="cm.aptoide.pt.dev"
```
Package name for Aptoide pointing to production environment is missing

mikieribeiro committed:
Added Package name for Aptoide pointing to production environment

Package name "cm.aptoide.pt" pointing to produciton (Aptoide) added

JoaoRaimundo commented on gradle.properties:
```diff
@@ -14,6 +14,8 @@ ROPSTEN_NETWORK_BACKEND_BASE_HOST=https://apichain-dev.blockchainds.com
 ROPSTEN_NETWORK_BACKEND_BASE_HOST_WS=https://api-dev.blockchainds.com
 
+APTOIDE_PACKAGE_NAME="cm.aptoide.pt"
+APTOIDE_PACKAGE_NAME_DEV="cm.aptoide.pt.dev"
```
Should we use this variable for dev build? Or just remove it since we don't use it?

JoaoRaimundo commented:
The notification should be dismissed when the user chooses to install the wallet.

mikieribeiro committed:
Updated gradle.properties and the buildDownloadButton() method

-removed package name "cm.aptoide.pt.dev" from gradle.properties

-The wallet install dialog is dismissed when the user clicks on the install button
